DRA: scheduler event handlers via assume cache

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Events that make pods scheduleable were triggered by the informer cache, not the assume cache. For "claim was deallocated", this led to a small, unlikely race if a pod got scheduled and stopped so quickly that the informer cache didn't ever see the "claim is allocated" state. The event handler now reacts to changes in the assume cache because that cache is guaranteed to receive the "claim is allocated" state which cause some pod to not get scheduled, because by definition the cache must have listed some other claim as using resources needed for that pod.

#### Which issue(s) this PR fixes:
Fixes #https://github.com/kubernetes/kubernetes/issues/123698

#### Does this PR introduce a user-facing change?
```release-note
DRA: fix some small, unlikely race condition during pod scheduling
```

/assign @kerthcet 

Do you have time to review?

/cc @towca 

This is related to the work that you are doing for the cluster autoscaler.

pohly committed:
scheduler: add FIFO queue

This is a basic implementation of a first-in-first-out queue with unbounded
size. It's useful for cases where a channel with fixed size might deadlock.

The caller is responsible for locking.

pohly committed:
scheduler: AddEventHandler for assume cache

This enables using the assume cache for cluster events.

pohly committed:
scheduler: central ResourceClaim assume cache

This enables connecting the event handler for ResourceClaim to the assume
cache, which addresses a theoretic race condition.

It may also be useful for implementing the autoscaler support, because now
the autoscaler can modify the content of the cache.

pohly committed:
ktesting: add Step

Step simplifies using WithStep because it creates a local scope where the same
tCtx variable is the one with the step name.

pohly commented:
/retest

kerthcet commented:
Will take a look later, sorry I didn't notice the assignment.

kerthcet commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -456,12 +458,11 @@ func addAllEventHandlers(
+				// populated. If a claim is not yet in the
+				// cache, scheduling will get retried once it
+				// is.
+				resourceClaimCache.AddEventHandler(buildEvtResHandler(at, framework.ResourceClaim, "ResourceClaim"))
```
We can make the`ResourceEventHandlerRegistration` part of the assumeCache and add to the handlerRegistration. Since the assumeCache is based on the informer, it makes sense to wait for the handler synced.

kerthcet reviewed (commented):
The eventQueue is a bit hard to understand, I may need more time on that.

kerthcet commented on pkg/scheduler/util/assumecache/assume_cache.go:
```diff
@@ -221,13 +249,26 @@ func (c *AssumeCache) add(obj interface{}) {
 		c.logger.Info("Error occurred while updating stored object", "err", err)
 	} else {
 		c.logger.V(10).Info("Adding object to assume cache", "description", c.description, "cacheKey", name, "assumeCache", obj)
+		for _, handler := range c.eventHandlers {
```
Seems we're maintaining two level event handlers here? I didn't quite get the idea why we need this? I read the comment of eventQueue, but still can't quite understand...

And do we have several handlers, what I learned from the PR is we only build the handler for `ResourceClaim`.

```
	for gvk, at := range gvkMap {		
                 case framework.ResourceClaim:
			if utilfeature.DefaultFeatureGate.Enabled(features.DynamicResourceAllocation) {
				// No need to wait for this cache to be
				// populated. If a claim is not yet in the
				// cache, scheduling will get retried once it
				// is.
				resourceClaimCache.AddEventHandler(buildEvtResHandler(at, framework.ResourceClaim, "ResourceClaim"))
			}
        }

pohly replied:
The chain of events is this now: claim informer -> assume cache -> plugin.

This ensures that local changes to the assume cache are properly reported to the plugin. This wasn't the case before, leading to the race explained in https://github.com/kubernetes/kubernetes/issues/123698

kerthcet replied:
Thanks for the explanation, I got the idea here, seems we have a MPSC architecture, we have producers like informer, local update, single consumer as the event handler. 

Based on that, maybe we'd better not consume the events immediately after producing them, but consuming them in another goroutine independently, that's say we'll move something like `emitEvents` to another place, decoupled with producer, which makes the code more readable and maintainable IMHO.

Moreover, we can hide the producer-consumer logic at cache layer, but locate them at the storage layer, take the update for example (this is just a demo) :

```
c.updateStore(objInfo) // rather than call c.store.Update() and push the events together.

func (c *AssumeCache) updateStore(oldObj, newObj) {
    c.store.Update()
    c.eventQueue.Push(oldObj, newObj)
}
```

My two cents.

pohly replied:
What you are proposing sounds *very* scary to me in terms of correctness and race conditions.

Can we keep the current architecture for event processing and just slightly extend it so that the source of events can be informers as well as local assume cache updates?

kerthcet replied:
I mean something like this https://github.com/kubernetes/kubernetes/commit/c7da77ca6350371a3e6f9977c420740438b30aa2, decouple the consumer logic.

pohly replied:
Can you explain why that is useful? What is the advantage of decoupling?

Event handlers are meant to complete quickly. That is already true for informer events. So making the code more complex to support long-running event handlers doesn't seem warranted to me.

kerthcet replied:
So the original idea is for simplify, the original code is a bit hard to understand, at least to me, so I don't think we're making it complex.

And I don't think it's slower, we'll consume the events continuously and we'll sent less events, on the contrary, I think we make less races, like 1) before when holding the lock, we'll iterate the handlers and send several related events, but now we only send one event, and 2) we'll call `emitEvents` at each OP, however, this will drain the whole events, which means we'll acquire unnecessary lock with several calls to `emitEvents` the same time.

BTW, for each OP, we'll send them to a queue and consume them from the queue in one function, what's your take here vs `invoking handlers directly`? Are you afraid the handlers will take a long time?

Would like to cc @alculquicondor @jsafrane for advices as well.

pohly replied:
Sorry, I am not following. The assume cache should emit the same number of events regardless in which goroutine the event handlers are called, so "now we only send one event" doesn't make sense to me.

What does "OP" stand for?

kerthcet replied:
What I mean is before, we loop the handlers to sent separated events like this during the lock time:
```
			for _, handler := range c.eventHandlers {
				handler := handler
				oldObj, obj := objInfo.latestObj, objInfo.apiObj
				c.eventQueue.Push(func() {
					handler.OnUpdate(oldObj, obj)
				})
			}
```

Now we do like this:
```
*Producer*: 
c.eventQueue.Push(event{OldObj: oldObj, Obj: obj})

*Consumer*:
for _, handler := range c.eventHandlers {
	handler := handler
	switch action {
	case "add":
		handler.OnAdd(event.Obj, false)
	case "update":
		handler.OnUpdate(event.OldObj, event.Obj)
	case "delete":
		handler.OnDelete(event.OldObj)
	}
}
``` 
So we're still doing the same thing, no handler will miss the event, but with different patterns.

> What does "OP" stand for?

Referring to the functions like `AddEventHandler`, `Restore`, `Assume`, `add/update/delete`.

pohly replied:
About "goroutines are more complex".

https://github.com/kubernetes/kubernetes/commit/c7da77ca6350371a3e6f9977c420740438b30aa2 currently has no "WaitForShutdown"-style API call, so unit tests cannot be sure that the goroutine has stopped. Lots of components in Kubernetes have that problem, but that doesn't mean that it's okay. When I write new code that spawns goroutines, I am making sure that cleanup and termination is handled. Can be added, of course.

The other drawback is that, as you noticed, unit tests are now no longer deterministic (`FIXME: Need to wait until consuming the events, it's a decoupled architecture.`). I'm curious how you want to fix that.

jsafrane replied:
I checked the code @pohly wrote, and while it's harder to follow the flow (informer -> assume cache -> plugin), I think it's usable.
I personally think c7da77c is adding only complications, without helping with code readability.

pohly commented on pkg/scheduler/eventhandlers_test.go:
```diff
@@ -447,8 +450,13 @@ func TestAddAllEventHandlers(t *testing.T) {
 
 			dynclient := dyfake.NewSimpleDynamicClient(scheme)
 			dynInformerFactory := dynamicinformer.NewDynamicSharedInformerFactory(dynclient, 0)
+			var resourceClaimCache *assumecache.AssumeCache
```
Good catch. I added new test cases which exercise the informer creation more completely.

The feature gate check in eventhandlers.go is a bit redundant: the plugin itself never asks for any of these events when it is disabled.

pohly commented on pkg/scheduler/framework/interface.go:
```diff
@@ -701,6 +702,10 @@ type Handle interface {
 	SharedInformerFactory() informers.SharedInformerFactory
 
+	// ResourceClaimInfos returns an assume cache of ResourceClaim objects
+	// which gets populated by the shared informer factory.
```
Added, with slightly different spelling.

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly committed:
DRA scheduler: use assume cache to list claims

This finishes the transition to the assume cache as source of truth for the
current set of claims.

The tests have to be adapted. It's not enough anymore to directly put objects
into the informer store because that doesn't change the assume cache
content. Instead, normal Create/Update calls and waiting for the cache update
are needed.

kerthcet reviewed (commented):
Are we going to have a bunch of resourceClaims, which may lead to the memory increase since the consumer side (move pods to scheduling queue) is slow?

pohly commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -456,12 +458,11 @@ func addAllEventHandlers(
-					return err
-				}
-				handlers = append(handlers, handlerRegistration)
+				// No need to wait for this cache to be
```
Okay, added.

pohly commented on pkg/scheduler/util/assumecache/assume_cache.go:
```diff
@@ -246,14 +287,29 @@ func (c *AssumeCache) delete(obj interface{}) {
 		c.logger.Error(err, "Failed to delete", "description", c.description, "cacheKey", name)
 	}
+
+	for _, handler := range c.eventHandlers {
```
I added `AssumeCache.pushEvent`.

pohly commented on pkg/scheduler/util/assumecache/assume_cache.go:
```diff
@@ -396,7 +466,53 @@ func (c *AssumeCache) Restore(objName string) {
+	c.rwMutex.Lock()
+	defer c.rwMutex.Unlock()
+
+	c.eventHandlers = append(c.eventHandlers, handler)
```
An event handler may get added at any time. That is true for informers and for the AssumeCache. Therefore we could have events already in the store.

kerthcet replied:
I saw it's true for several tests that informer will add objects in prior, but is that true for the running code, because  `When building event handler, the informer hasn't started yet`, just for confirmation.

pohly replied:
The assume cache itself doesn't care whether the informer is running or not. It just needs to do the right thing based on what it currently has, which may or may not be an empty store.

pohly commented on pkg/scheduler/util/assumecache/assume_cache.go:
```diff
@@ -396,7 +466,53 @@ func (c *AssumeCache) Restore(objName string) {
+// in which they were stored there (FIFO).
+func (c *AssumeCache) emitEvents() {
+	for {
+		c.rwMutex.Lock()
```
I think that's premature optimization. This is being used for resource claims with a single handler, so I don't expect many concurrent events.

As part of the work for 1.31 I will run scheduler_perf again to look for CPU hotspots. We can still optimize this later if it really is a problem.

kerthcet replied:
yeah, for sure.

pohly commented on pkg/scheduler/util/assumecache/assume_cache_test.go:
```diff
@@ -118,6 +120,79 @@ func verifyList(tCtx ktesting.TContext, assumeCache *AssumeCache, expectedObjs [
+		InitialList: initialList,
+	})
+
+	if m.cache != nil {
```
This checks that `m.cache.List` can be called while that cache is delivering the event.

kerthcet replied:
Trick then, new skills learned.

pohly commented on pkg/scheduler/util/assumecache/assume_cache_test.go:
```diff
@@ -226,27 +325,161 @@ func TestEvents(t *testing.T) {
+	})
+	ktesting.Step(tCtx, "nil add", func(tCtx ktesting.TContext) {
+		informer.add(nil)
+		verify(tCtx, cache, key, newObj, newObj)
```
No. Added.

pohly commented on pkg/scheduler/util/assumecache/assume_cache_test.go:
```diff
@@ -226,27 +325,161 @@ func TestEvents(t *testing.T) {
+	tCtx, cache, informer := newTest(t)
+	handlers := make([]mockEventHandler, 5)
+
+	objs := make([]metav1.Object, 0, 20)
```
I've simplified by removing the pre-allocation.

pohly commented on pkg/scheduler/util/queue/fifo.go:
```diff
@@ -0,0 +1,95 @@
+// The null FIFO is a valid empty queue.
+//
+// Access must be protected by the caller when used concurrently by
+// differerent goroutines, the queue itself implements no locking.
```
Fixed.

pohly commented on pkg/scheduler/util/queue/fifo.go:
```diff
@@ -0,0 +1,95 @@
+	q.start++
+	if q.start == len(q.elements) {
+		// Wrap around.
+		q.start = 0
```
Added, together with a test.

kerthcet replied:
Also update the number 4 to `normalSize` in L62.

pohly replied:
Updated.

pohly commented:
@kerthcet: I hope [the last push](https://github.com/kubernetes/kubernetes/compare/6914689aa51db8c163d151a3b32e67a7b78f4e5c..f61703d70e74434681bd3f8a62ad1a28e57b8808) addresses your feedback.

pohly commented:
Weird. How can this PR break `TestFrameworkHandler_IterateOverWaitingPods/pods_with_same_profile_are_waiting_on_permit_stage`?

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124595/pull-kubernetes-unit/1804178145381715968

Let's retest, just to see whether it is consistent.

/retest

pohly commented:
> How can this PR break `TestFrameworkHandler_IterateOverWaitingPods/pods_with_same_profile_are_waiting_on_permit_stage`?
> 
> https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124595/pull-kubernetes-unit/1804178145381715968

Looks like a flake, it also occurs in periodic jobs from time to time.

I filed a bug: https://github.com/kubernetes/kubernetes/issues/125653

kerthcet commented on pkg/scheduler/util/assumecache/assume_cache_test.go:
```diff
@@ -226,27 +325,161 @@ func TestEvents(t *testing.T) {
+	var wg sync.WaitGroup
+
+	for i := range handlers {
+		handlers[i].block = tCancelCtx.Done()
```
second ping this.

pohly replied:
It causes the event handlers to block. The comment above explains that intent.

The corresponding code is in `mockEventHandler.OnAdd`:
```
	if m.block != nil {
		<-m.block
	}
```

This blocking of the handlers is necessary to force a certain execution order.

kerthcet commented:
/lgtm
/approve

Thanks @pohly for the patience!

pohly commented:
/retest

pohly commented:
/retest
